### ISSUE TITLE
[v6.0]: Roadmap for v6.0 release

### ISSUE THREAD

**Issue by Cielquan** [rstcheck/rstcheck#104]
# Roadmap for v6.0 release

You can also see the work that need to be done until the final release in the [v6.0.0 milestone](https://github.com/myint/rstcheck/milestone/1).

Feedback on the pre-releases can be posted here and is appreciated.

## `v6.0.0alpha1`

- rstcheck/rstcheck#95

## `v6.0.0rc1`

- rstcheck/rstcheck#99
    Adding back a logging *(to console)* system, which was dropped in rstcheck/rstcheck#100

- rstcheck/rstcheck#102
    Including:
    - rstcheck/rstcheck#97
    - rstcheck/rstcheck#98

## `v6.0.0` (final)

No additions to rc1 planned.
Before the final release a Hosting solution must be found for the docs. I am onto something.

## Post `v6.0.0`

The addition of new features will be postponed for after `v6.0.0`.
Some features I then plan to address are (this list is not complete):
- rstcheck/rstcheck#103 
- rstcheck/rstcheck-core#8
- rstcheck/rstcheck-core#5

**Comment by Cielquan** [comment]
v6.0.0a1 got released

**Comment by zaneselvans** [comment]
We're using `rstcheck` as a pre-commit hook in several of our repos, integrated with https://pre-commit.ci and the `v6.0.0a1`  release has bumped the version of the hooks, and seems to be breaking the pre-commit.ci runs, see [this pre-commit.ci autoupdate PR](https://github.com/catalyst-cooperative/intake-sqlite/pull/8).

In our `.pre-commit.yaml` the `rstcheck` section looks like:
```yaml
- repo: https://github.com/myint/rstcheck
  rev: v6.0.0a1
  hooks:
  - id: rstcheck
    additional_dependencies: [sphinx]
    args: ['--config', 'tox.ini']
```

The configuration in `tox.ini` is:
```ini
[rstcheck]
report = warning
ignore_roles =
    pr,
    issue,
    user
ignore_messages = (Hyperlink target ".*" is not referenced\.$)
ignore_directives =
    bibliography,
```

Running the pre-commit hook at the command line like this succeeds just fine:
```sh
pre-commit run --all-files rstcheck
```

But running `rstcheck` directly like this (in a Python environment with the `sphinx` dependencies installed as well):
```sh
rstcheck --recursive --config tox.ini ./
```

Results in [the same failure that is showing up on pre-commit.ci](https://results.pre-commit.ci/run/github/490949616/1652735390.TMfYgMsWT8K89tOj07toFQ):
```txt
WARNING: while setting up extension sphinx.addnodes: node class 'toctree' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_signature' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_signature_line' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_content' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_inline' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_name' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_addname' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_type' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_returns' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_parameterlist' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_parameter' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_optional' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_annotation' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_space' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_name' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_operator' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_punctuation' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_keyword' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_keyword_type' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_literal_number' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_literal_string' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'desc_sig_literal_char' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'versionmodified' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'seealso' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'productionlist' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'production' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'index' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'centered' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'acks' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'hlist' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'hlistcol' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'compact_paragraph' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'glossary' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'only' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'start_of_file' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'highlightlang' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'tabular_col_spec' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'pending_xref' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'number_reference' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'download_reference' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'literal_emphasis' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'literal_strong' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'manpage' is already registered, its visitors will be overridden
WARNING: while setting up extension sphinx.addnodes: node class 'meta' is already registered, its visitors will be overridden
Success! No issues detected.
```

Downgrading with `pip install rstcheck[sphinx]==5.0.0` and re-running the above command results in no output at all.

Is this an expected behavior?

**Comment by Cielquan** [comment]
Yes and No. Version 6 has a lot of breaking changes and for the final Release I will create a migration guide with the new documentation.

Let me go through your post.

The pre-commit config looks fine to me and should not have changed at all.

The rstcheck config on the other hand got some breaking changes see the [changelog](https://github.com/myint/rstcheck#a1-2022-05-13) for more information.
The concrete errors I can spot are
- `report` was renamed to `report_level`
- Multiline strings are no longer supported because of changes in the code base. But I guess it would be a good idea to add the support for multiline strings back. I will take a look at the matter.

Because of the new config handling additional config options which are set in the `[rstcheck]` section are just ignored. Therefore `report` did not show any error and the default level is used. I will take a look at that matter too. Maybe a `--strict-config` flag or so would be good, which would error on additional config options which are not supported.

I saw the behavior with `WARNING: while setting up extension sphinx.addnodes: node class ...` in failing tests with pytest too. But I, until now, never saw it on the actual CLI. But this was an issue on my ToDo list to investigate later.

I guess the local pre-commit run did not print those warnings because the exit code is 0, because no linting issues were found, and on 0 pre-commit does not print any output by default AFAIK. rstcheck exists 1 if any linting issue is found or an unhandled exception is raised.

Like stated above the printing of the warnings on the local direct run is new to me, but I will test that. And here you can see, that `Success! No issues detected.` is printed at the end, which is only printed if no issues were found and then exists 0. [See the source](https://github.com/myint/rstcheck/blob/master/src/rstcheck/runner.py#L145-L147).

The pre-commit.ci run you linked on the other hand fails, which is strange. Unfortunately the output is truncated and we cannot see if the success message is printed or what the actual issue is.

Thanks for the feedback.

**Comment by zaneselvans** [comment]
I think it would be good to consider adding multiline strings back in for two reasons:
* Having each element of a list on a separate line makes it clearer exactly what has changed in diffs / PRs.
* I assumed that acceptable formatting of elements in the ini file were determined by the file format, rather than potentially being treated differently by every tool getting its settings from the file.

Getting a warning about additional configuration options that aren't supported would be great. It can be very confusing & frustrating when things mysteriously stop behaving as expected without making any noise about something having changed.

**Comment by Cielquan** [comment]
Multi-line string supported will be release with v6.0.0a2. I will make a release when #113 is fixed.

**Comment by Cielquan** [comment]
v6.0.0a2 is live

**Comment by Cielquan** [comment]
The contents for the in the OP planned beta and final releases are both ready for review/merging.

Therefore the beta release will be skipped and a release candidate will be released after merging.

**Comment by Cielquan** [comment]
v6.0.0rc1 is live

If no bugs are found this is the last release before the final release.

**Comment by DanielNoord** [comment]
Congratulations on almost releasing the new release!

Just want to provide one more data point. Issues related to https://github.com/myint/rstcheck/issues/113 have indeed been fixed in the `rc`. However, `rstcheck` now seems to check more code blocks than it did before. While this is obviously a good thing it also makes it somewhat unusable for us over at `pylint`.
We have code that produces `SyntaxErrors` in our changelog to show for which code we can warn when we introduce new checks. I looked in the documentation but there doesn't seem to be a way to ignore specific code blocks. I guess that's what 
#103 is about. Other users might run into similar issues I guess. Not sure if this would be a release blocker for you, but implementation of #103 would be much appreciated 😄
If you would like me to test anything, feel free to ping!

Relevant PR: https://github.com/PyCQA/pylint/pull/6631.

**Comment by Cielquan** [comment]
Thanks :)

---

Well, this behavior really surprises me and I cannot currently tell what the reason is. I opened a new issue to further investigate the issue.

Yes, this is a hard blocker for me. Even so there are a lot of breaking changes, I want the user's to be able to use rstcheck after a migration, almost *like before*.

---

The new documentation is live at RTD: http://rstcheck.rtfd.io/

**Comment by Cielquan** [comment]
Because of some bugs the final release will be postponed by a bit.

**Comment by Cielquan** [comment]
With the current state I split this project into a core libraray `rstcheck-core` and a cli application `rstcheck`.

`rstcheck-core` was just released with v1.0.0rc1 and is used in v6.0.0rc3, which was also just release, of `rstcheck`.

The split should reduce the amount of breaking releases in the application part because of changes in library's API, which a CLI user does not care about.

Therefore I currently plan to release the final v6 version next weekend.
`rstcheck-core` will probably have its 1.0.0 release soon too, but may also soon have the first breaking release because of #90 which is a pretty deep issue.
